safe/impact_functions/inundation/flood_raster_road_qgis_gdal/impact_function.py

```diff
@@ -13,16 +22,116 @@
+        which are within threshold (thr_min < V < thr_max) """
+
+    rp = flood.dataProvider()
+    rb = rp.block(1, rp.extent(), rp.xSize(), rp.ySize())
+    rp_xmin = rp.extent().xMinimum()
+    rp_ymax = rp.extent().yMaximum()
+    cellx = rp.extent().width() / rp.xSize()
+    celly = rp.extent().height() / rp.ySize()
+
+    vl = QgsVectorLayer("Polygon?crs=epsg:4326", "flood polygons", "memory")
+    feats = []
+
+    for y in xrange(rp.ySize()):
+        for x in xrange(rp.xSize()):
+            v = rb.value(y,x)
+            if v < thr_min or v > thr_max: continue
+            x0,x1 = rp_xmin+(x*cellx), rp_xmin+((x+1)*cellx)
+            y0,y1 = rp_ymax-(y*celly), rp_ymax-((y+1)*celly)
+            outer_ring = [QgsPoint(x0,y0),QgsPoint(x1,y0),QgsPoint(x1,y1),QgsPoint(x0,y1),QgsPoint(x0,y0)]
+            f = QgsFeature()
+            f.setGeometry(QgsGeometry.fromPolygon([outer_ring]))
+            feats.append(f)
+
+    res, feats = vl.dataProvider().addFeatures(feats)
+
+    flood_cells_map = {}
+    for f in feats:
+        flood_cells_map[f.id()] = f
+
+    if QGis.QGIS_VERSION_INT >= 20800:
+        # woohoo we can use bulk insert (much faster)
+        index = QgsSpatialIndex(vl.getFeatures())
+    else:
+        index = QgsSpatialIndex()
+        for f in vl.getFeatures():
+            index.insertFeature(f)
+
+    return index, flood_cells_map
+
+
+def _add_output_feature(feats, geom, is_flooded, fields, orig_attrs, target_field):
+    # make sure to explode multi-part features
+    geoms = geom.asGeometryCollection() if geom.isMultipart() else [ geom ]
+    for g in geoms:
+        f = QgsFeature(fields)
+        f.setGeometry(g)
+        for attr_no, attr_val in enumerate(orig_attrs):
+            f.setAttribute(attr_no, attr_val)
+        f.setAttribute(target_field, is_flooded)
+        feats.append(f)
+
+
+
+def _union_geometries(geoms):
+    """ Return a geometry which is union of the passed list of geometries """
+    if QGis.QGIS_VERSION_INT >= 20400:
+        # woohoo we can use fast union (needs GEOS >= 3.3)
+        return QgsGeometry.unaryUnion(geoms)
+    else:
+        # uhh we need to use slow iterative union
+        if len(geoms) == 0:
+            return QgsGeometry()
+        result_geometry = QgsGeometry(geoms[0])
+        for g in geoms[1:]:
+            result_geometry = result_geometry.combine(g)
+        return result_geometry
+
+
+def _intersect_roads_flood(roads, request, index, flood_cells_map, output_layer, target_field):
+
+    feats = []
+    fields = output_layer.dataProvider().fields()
+
+    rd = 0
+    for f in roads.getFeatures(request):
+        ids = index.intersects(f.geometry().boundingBox())
+        geoms = [ flood_cells_map[i].geometry() for i in ids ]
+        flood_geom = _union_geometries(geoms)
+
+        in_geom = f.geometry().intersection(flood_geom)
+        if in_geom and (in_geom.wkbType() == QGis.WKBLineString or in_geom.wkbType() == QGis.WKBMultiLineString):
```

Comment:
Could you add a few comments here explaining the logic so that if someone finds the code later it is clear for them?